packages/core-sdk/resolvers/token-resolvers/resolvers/bancor/src/query/functions/getTokenComponents.ts

```diff
@@ -0,0 +1,141 @@
+    args: null,
+    connection: connection,
+  });
+  if (tokenAddressesRes.isErr) {
+    throw new Error("Invalid protocol token");
+  }
+  return tokenAddressesRes.unwrap().split(",");
+}
+
+export function getTokenComponents(input: Input_getTokenComponents): Interface_TokenComponent {
+  if (env == null) throw new Error("env is not set");
+  const connection = (env as QueryEnv).connection;
+
+  const token = Token_Query.getToken({
+    address: input.tokenAddress,
+    m_type: Token_TokenType.ERC20,
+  }).unwrap();
+
+  if (!token) {
+    throw new Error(`Token ${input.tokenAddress} is not a valid ERC20 token`);
+  }
+
+  const converterAddress: string = getConverterAddress(token.address, connection);
+  const poolTokenAddresses: string[] = getPoolTokenAddresses(converterAddress, connection);
+
+  const tokenDecimals: string = BigInt.fromUInt16(10).pow(token.decimals).toString();
+  const totalSupply: Big = Big.of(token.totalSupply.toString()) / Big.of(tokenDecimals);
+
+  const components: Interface_TokenComponent[] = [];
+  let unresolvedComponents: i32 = 0;
+
+  for (let j = 0; j < poolTokenAddresses.length; j++) {
+    let adjBalance: Big;
+    const underlyingTokenAddress: string = poolTokenAddresses[j];
+    if (underlyingTokenAddress.toLowerCase() == ETH_ADDRESS) {
+      // Eth special case
+      const balanceRes = Ethereum_Query.callContractView({
+        connection: connection,
+        address: converterAddress,
+        method: "function reserveBalance(address reserveToken) public view returns (uint256)",
+        args: [underlyingTokenAddress],
+      });
+      if (balanceRes.isErr) {
+        unresolvedComponents++;
+        continue;
+      }
+      const balance: string = balanceRes.unwrap();
+      const underlyIngDecimals: string = BigInt.fromUInt16(10).pow(18).toString();
+      adjBalance = Big.of(balance) / Big.of(underlyIngDecimals);
+    } else {
+      // get underlying token
+      const underlyingToken: Interface_Token = changetype<Interface_Token>(
+        Token_Query.getToken({
+          address: underlyingTokenAddress,
+          m_type: Token_TokenType.ERC20,
+        }).unwrap(),
+      );
+      if (!underlyingToken) {
+        unresolvedComponents++;
+        continue;
+      }
+
+      // get underlying token balance
+      const balanceRes = Ethereum_Query.callContractView({
+        connection: connection,
+        address: underlyingToken.address,
+        method: "function balanceOf(address account) public view returns (uint256)",
+        args: [converterAddress],
+      });
+      if (balanceRes.isErr) {
+        unresolvedComponents++;
+        continue;
+      }
+      const balance: string = balanceRes.unwrap();
+      const underlyIngDecimals = BigInt.fromUInt16(10).pow(underlyingToken.decimals).toString();
+      adjBalance = Big.of(balance) / Big.of(underlyIngDecimals);
+    }
+
+    // calculate and push rate
+    const rate = (adjBalance / totalSupply).toString();
```

Comment:
At first I updated everything so the implementations match those from Zerion. But I was confused about some things, so I spoke with a dev at Bancor in discord. He told me there are no v2 pools deployed, and that v2.1 doesn't put liquidity tokens in people's wallets. The pool liquidity tokens are all held by the protocol itself. So actually we only need a v1 resolver. 

The reason we were getting the correct answer with my original implementation is because Zerion's method--calling the v1 contract--uses the same formula that I had originally implemented to calculate the rate. I'm going to clean up everything and refactor soon, and use that original formula (which matches the v1 contract method).